Moebius Transform Operator and Tustin's method for LTI-Systems

artpelling opened the pull request:
I have this code here that I would like to integrate in order to have Bilinear Transformations of LTI-systems similarly to Matlabs `c2d` and `d2c` commands. So far I have implemented Tustin's method and I am looking to extend this to higher order approximations in the future. 

I am particularly interested in arbitrary Moebius substitutions of the frequency argument (not only the bilinear transformation) which is why I chose a more functional approach and implemented a `MoebiusTransform` operator directly. Maybe this can be used elsewhere in the future (I have added the CayleyTransform as an inspiration).

I would love to get some feedback, IMO the method `LTIModel.moebius_substitution` could use some improvement in both naming and the code. The method substitutes the frequency argument `omega` by a MoebiusTransform (please see https://ieeexplore.ieee.org/document/489281/ for reference). I feel like I am not using pyMORs structure optimally in the current implementation.

The docstrings are sparse and sketchy, I will obviously add proper ones and tests at a later stage.

pmli commented on src/pymor/models/iosys.py:
```diff
@@ -928,6 +929,45 @@ def get_ast_spectrum(self, ast_pole_data=None, mu=None):
+            source_id=self.A.source_id,
+            range_id=self.A.range_id)
+
+        # At = to_matrix(Gamma @ (d * self.A - b * self.E))
+        At = M.apply_inverse(M.source.from_numpy(self.A.matrix)).to_numpy().T
+        Bt = to_matrix(np.sign(s) * v * Gamma @ self.B)
+        Ct = to_matrix(v * self.C @ Gamma)
+        Dt = to_matrix(self.D + c * self.C @ Gamma @ self.B)
```
`to_matrix` should be avoided, unless this is expected to be used only for small models.

artpelling commented on src/pymor/operators/numpy.py:
```diff
@@ -531,3 +531,111 @@ def H(self):
         )
+
+
+class MoebiusTransform(NumpyGenericOperator):
+    """A Moebius transform operator.  Maps complex square matrices to complex square matrices.
```
Yes, I agree, but it does not seem to fit into the algorithms or tools modules either... Since the Moebius transfrom maps square matrices to square matrices and VectorArrays are (ab)used as matrices in other places in pyMOR, I went with an `Operator` implementation that checks that `len(U) == U.dim` in `MoebiusTransform.apply(U)`.

Does anyone have a better idea?

pmli replied:
Maybe it fits in the `models` module?
Also, maybe it's better to think of the Moebius transform just as a rational function, without `apply` and `apply_inverse` methods.

artpelling replied:
> Maybe it fits in the `models` module?

That would be possible. It would actually be nice, if it could take and return `Operators`. Do you think there is merit in implementing other operator -> operator mappings in a similar way? 

> Also, maybe it's better to think of the Moebius transform just as a rational function, without `apply` and `apply_inverse` methods.

I like the fact that they are objects, because it allows for different constructors and the group structure is reflected by the `matmul` method, so that multiple transforms can be combined into one. Could this be achieved with just a function?

pmli replied:
I agree, I meant implementing `MoebiusTransform` as a class, but without `Operator` semantics (I guess the `coefficients` attribute and the `__matmul__` method would be enough). This is because I don't see what you gain by treating it as a mapping from `Operators` to `Operators`, besides a way to write the transformed `A`. You would still transform `B`, `C`, and `D` by hand, so why not also `A`.

artpelling replied:
I intended to make the implementation as general as possible such that it could be used elsewhere in the future. I totally agree, that this is not needed in the `LTIModel` and so I am also happy to move a scalar version to the models module.

artpelling replied:
I just starting doing that, but I realized that some kind of `apply` is still needed in case one wants to actually apply the Moebius transform to some frequencies. I'm fine with only having a scalar implementation though, so it does not necessarily need to work on `VectorArrays`. Should I still call the method `apply` or is it better to chose a different approach?

pmli replied:
What do you mean by an `apply` method here? A method that takes `s` and returns `(a*s+b)/(c*s+d)`? You could use `__call__` to do that if you'd like.

artpelling replied:
okay, great. I have moved it over to `models.transforms` and inherited from `Model`. Let me know if that is what you had in mind!

artpelling commented on src/pymor/models/iosys.py:
```diff
@@ -928,6 +929,45 @@ def get_ast_spectrum(self, ast_pole_data=None, mu=None):
+        s = a * d - b * c
+        v = np.sqrt(np.abs(s))
+        Gamma = NumpyMatrixOperator(
+            spla.inv(to_matrix(a * self.E - c * self.A)),
```
I agree, however, using `InverseOperator` leads to a problem in `LTIModel.transfer_function.tf` where `to_matrix` is called on `D`, because there is no `MatchingRule` for `InverseOperator`...

pmli replied:
I see. Then it would probably be good for `to_matrix` not to recurse through `ConcatinationOperator`, but instead use something like `concat_op.as_range_array(mu=mu).to_numpy().T` (that way, the `InverseOperator` should not be turned to a matrix).

artpelling replied:
Good idea. I did some small checks and the speeds of both variants are somewhat equal. However, one can encounter problems with large operators, because `as_array_max_length()` defaults to 100. Not a big issue I suspect, because it can be easily raised if needed.

pmli commented on src/pymor/algorithms/to_matrix.py:
```diff
@@ -133,13 +137,7 @@ def action_ComponentProjectionOperator(self, op):
-            else:
-                mats = mats[:-2] + [mats[-2].dot(mats[-1])]
-        return mats[0]
+        return op.as_range_array().to_numpy().T
```
```suggestion
        return op.as_range_array(mu=self.mu).to_numpy().T
```

Also, I see now that it might be good to keep the old implementation in case the result is a sparse matrix. Something like the following should work:

```python
try:
    mats = [self.apply(o) for o in op.operators]
    from functools import reduce
    from operator import matmul
    return reduce(matmul, mats)
except NoMatchingRuleError:
    return op.as_range_array(mu=self.mu).to_numpy().T
```

artpelling commented on src/pymor/models/iosys.py:
```diff
@@ -928,6 +930,43 @@ def get_ast_spectrum(self, ast_pole_data=None, mu=None):
+        s = a * d - b * c
+        v = np.sqrt(np.abs(s))
+        Gamma = InverseOperator(a * self.E - c * self.A)
+        At = NumpyMatrixOperator(
+            M.apply_inverse(M.source.from_numpy(self.A.matrix)).to_numpy().T,
+            source_id=self.A.source_id,
+            range_id=self.A.range_id)
```
I have added a basic implementation of `assemble_concat` in #1618. Maybe something similar could be done for the inverse of a `NumpyMatrixOperator`?

pmli replied:
How about moving `Gamma` to the `E` matrix by state-space transformation? It looks like instead using

```python
Et = a * self.E - c * self.A
At = a * self.A - b * self.E
Ct = v * self.C
```

defines an equivalent system.

Also, it would be good to allow for `self.E` to not necessarily be an `IdentityOperator`.

artpelling replied:
Hmm. I'm not sure I get what you mean. I can construct a system with
```
        Et = d * self.E + c * self.A
        At = (a * self.A + b * self.E)
        Bt = np.sign(s) * v * self.B 
        Ct = v * self.C @ InverseOperator(Et)
        Dt = self.D - c * self.C @ InverseOperator(Et) @ self.B
```
But that still leaves the inverse in `Ct` and `Dt`...

artpelling replied:
I don't fully understand what is going on but it seems like having the inverses just in `Ct` and `Dt` solves the problem with the slow transfer function evaluations. So I guess it works out like this, thanks for the suggestion!

pmli replied:
The speed improvement is good. It is expected because the previous implementation involved solving a linear system with a lot of right-hand sides (basically, `Et.apply_inverse(At)`).

> > Also, it would be good to allow for self.E to not necessarily be an IdentityOperator.
> 
> Yes definitely. So far, I only have that one source that does things in absence of `E`. I'm not sure if/how these results can be generalized at the moment.

I guess that (using pencil and paper) you could apply the Moebius transform to (E<sup>-1</sup> A, E<sup>-1</sup> B, C), and then see if you can get rid of E<sup>-1</sup> in the transformed system. This should then generalize the procedure to systems where E is not the identity.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,107 @@
+import numpy as np
+from scipy.linalg import inv, null_space
+
+from pymor.models.interface import Model
+
+
+class MoebiusTransform(Model):
```
```suggestion
from pymor.core.base import ImmutableObject


class MoebiusTransform(ImmutableObject):
```

It is not really a model, similar to [`TransferFunction`](https://github.com/pymor/pymor/blob/b5eb4e4b27f5a2028776e401e1813792e2044d89/src/pymor/models/transfer_function.py#L15).

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,107 @@
+        elif c == 0 and np.isinf(x).any():
+            return np.inf * I
+        else:
+            return (a * x + b * I) @ inv(c * x + d * I)
```
It would be good to use `solve` instead of `inv`.

artpelling replied:
I went back to supporting just scalars, because I realised that a thorough implementation of matrix maps would need more and there is no use case at the moment. I went ahead to allow one-dimensional arrays or lists as function inputs so that frequency vectors can be mapped.

artpelling commented:
I have somehow messed up the history again. @pmli if you are okay with the current state, I can rebase on main, squash everything into one commit and convert this to a non-draft PR.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,90 @@
+        self.__auto_init(locals())
+
+    @classmethod
+    def from_points(cls, w, z=[0, 1, np.inf], name=None):
```
It is good to avoid using mutable defaults.

```suggestion
    def from_points(cls, w, z=(0, 1, np.inf), name=None):
```

pmli commented:
> I have somehow messed up the history again. @pmli if you are okay with the current state, I can rebase on main, squash everything into one commit and convert this to a non-draft PR.

Rebasing would also make it easier to review. You don't have to squash it into one commit, e.g., using one commit per file changed would make the git history nicer (also, we often use git commit messages of the form `[module] short message`, but we do not really enforce it).

pmli commented on src/pymor/algorithms/to_matrix.py:
```diff
@@ -47,6 +49,10 @@ def __init__(self, format, mu):
         super().__init__()
         self.__auto_init(locals())
 
+    @match_class(MoebiusTransform)
+    def action_MoebiusTransform(self, op):
+        return op.coefficients.reshape(2, 2)
+
```
I guess this can be removed?

artpelling replied:
probably yes!

pmli reviewed (commented):
It looks to me to be in a good shape. What is left are docstrings and some unit tests.

artpelling commented:
I renamed `MoebiusTransform` to `MoebiusTransformation`, because apparently the former refers to Möbius' inversion formula. This leads to inconsistent but correct naming of the derived classes.

pmli commented on src/pymor/models/iosys.py:
```diff
@@ -939,6 +942,102 @@ def get_ast_spectrum(self, ast_pole_data=None, mu=None):
+            Continuous-time |LTIModel|.
+        """
+        if method != 'Tustin':
+            return NotImplemented
```
```suggestion
            return NotImplementedError
```

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,156 @@
+
+    def __init__(self, coefficients, normalize=False, name=None):
+        assert len(coefficients) == 4
+        coefficients = np.array(coefficients, dtype=complex)
```
What would be the reason to convert to `complex`? (Note that NumPy has `np.complex64`, `np.complex128`, and `np.complex256`.)

artpelling replied:
This is not needed, you are right. I was afraid that the old Python behaviour where int divisions return ints not floats could lead to problems in `__call__`... 👻

artpelling replied:
This is actually needed to resolve the `TypeError` that results from `/=`. Not sure exactly why though...

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,156 @@
+            A tuple, list or |NumPy array| of three complex numbers represent the images of `w`.
+            Defaults to `(0, 1, np.inf)`.
+        name
+            Name of the transformation.
+        """
```
"Returns" part is missing.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,156 @@
+        Parameters
+        ----------
+        normalize
+            If `True`, the coefficients are normalized, i.e. :math:`ad-bc=1`. Defaults to `False`.
+        """
```
"Returns" part is missing.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,156 @@
+    def __init__(self, x, name=None):
+        assert x > 0
+        super().__init__([x, -x, 1, 1], name=name)
+        self.__auto_init(locals())
```
`__auto_init` would make `x` an attribute. Is this necessary, or is it something that is good to have?

artpelling replied:
I don't know that it will ever be actually useful, but it is the one parameter that defines the transformation. So it seemed appropriate to have it as an attribute.

pmli replied:
Ok, sounds good.

pmli commented:
It would be good to add tests for Moebius transformations and maybe also for `to_discrete` and `to_continuous` methods.

Also, now that it's settled that these transformations are not `Operators`, maybe `__mul__` instead of `__matmul__` is better? Or would that be confused with the elementwise product of functions?

renefritze commented:
I've add you to the safelist now @artpelling , so this message should disappear with the next push/commit.

artpelling commented:
OK, great thank you!

artpelling commented:
> I've add you to the safelist now @artpelling , so this message should disappear with the next push/commit.

:thinking: it is still there. don't know what I can do about it. It keeps happening when I rebase on main for some weeks now

renefritze commented:
I fixed the bot for real now, as demonstrated by the gitlab build running and the deleted comment not re-appearing.

artpelling commented:
> I fixed the bot for real now, as demonstrated by the gitlab build running and the deleted comment not re-appearing.

Nice. Now there is only the issue that some tests are failing. I am pretty sure this has nothing to do with the PR itself, because the same tests also fail elsewhere. Just in case you are not aware of that already..

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+# This file is part of the pyMOR project (http://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
```
```suggestion
# This file is part of the pyMOR project (https://www.pymor.org).
# Copyright pyMOR developers and contributors. All rights reserved.
# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
```

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+    """
+
+    def __init__(self, coefficients, normalize=False, name=None):
+        assert len(coefficients) == 4
+        coefficients = np.array(coefficients)
```
```suggestion
        coefficients = np.array(coefficients)
        assert coefficients.shape == (4,)
```

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+        assert coefficients[0]*coefficients[3] != coefficients[1]*coefficients[2]
+
+        if normalize:
+            coefficients = coefficients / np.sqrt(coefficients[0]*coefficients[3]-coefficients[1]*coefficients[2])
```
```suggestion
            coefficients /= np.sqrt(coefficients[0]*coefficients[3]-coefficients[1]*coefficients[2])
```

artpelling replied:
Hmm, I had to change this line because your suggestion gave me the `TypeError`


` TypeError: ufunc 'true_divide' output (typecode 'd') could not be coerced to provided output parameter (typecode 'l') according to the casting rule ''same_kind''`

When executing the tests.

artpelling replied:
I like the idea. As far as I can tell, the normalization of coefficients is not too common anyways, so I would say its fine to have our own interpretation. I have put the absolute value inside the square root. This changes the result for complex values but it is still valid.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+    coefficients
+        A tuple, list or |NumPy array| containing the four coefficients `a,b,c,d`.
+    normalize
+        If `True`, the coefficients are normalized, i.e. :math:`ad-bc=1`. Defaults to `False`.
```
```suggestion
        If `True`, the coefficients are normalized, i.e., :math:`ad-bc=1`. Defaults to `False`.
```

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+        Parameters
+        ----------
+        normalize
+            If `True`, the coefficients are normalized, i.e. :math:`ad-bc=1`. Defaults to `False`.
```
```suggestion
            If `True`, the coefficients are normalized, i.e., :math:`ad-bc=1`. Defaults to `False`.
```

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+        return (
+            f'{self.name}: C^1 --> C^1\n'
+            f'       ({self.coefficients[0]:.1f})*z + ({self.coefficients[1]:.1f})\n'
+            'f(z) = --------------------------------\n'
```
It would be good to make the line depend on the length of the numerator and denominator.

pmli commented on src/pymor/models/transforms.py:
```diff
@@ -0,0 +1,171 @@
+
+    def __str__(self):
+        return (
+            f'{self.name}: C^1 --> C^1\n'
```
Maybe Unicode works...

```suggestion
            f'{self.name}: ℂ --> ℂ\n'
```

artpelling commented:
@renefritze gitlab is failing again :question:

artpelling commented:
@pmli The line of the fraction now depends on the length of the numerator and denominator in the `__str__` method. I have added a casting to float in case integer values are provided with `normalize=True`.

Can anyone explain why `|MoebiusTransformation|` in the docstrings does not work but e.g. `|Operator|` does? I have changed to :class: syntax instead...

sdrave commented:
@artpelling, the problem with the failing test was that you are not allowed to call `as_range_array` on an `Operator` which has something different than a `NumpyVectorSpace` as source. I have moved the fallback in `to_matrix` to an additional `as_array` action. As this should also work for the children in the `Concatenation`, I do not think it is necessary to catch `NoMatchingRuleError`.

Further I have reverted your streamlined version of the matrix-matrix products as it seems to neglect one optimization for sparse-dense products.
